### PR TITLE
feat(http): export a list of http status codes

### PR DESCRIPTION
**Pull request by QuentinFchx**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: https://github.com/angular/angular/issues/23543

## What is the new behavior?
The `@angular/common/http` now exports a list of http status codes.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
I would need directions if this needs to be added to the documentation.

### CODE REVIEW

File: packages/common/http/src/status_codes.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/**
+ * Http status codes.
+ */
+export enum HttpStatusCode {
```
Comment (AhnpGit) [inline packages/common/http/src/status_codes.ts]: I think it should be const enum like the discussion in #23543

**Review by ahnpnl — CHANGES_REQUESTED** [review]

File: packages/common/http/src/status_codes.ts (hunk 1 above)
Comment (QuentinFchx) [reply packages/common/http/src/status_codes.ts]: Yup, my bad. I wanted to be consistent with existing enums declaration (such as [ChangeDetectionStrategy](https://github.com/angular/angular/blob/master/packages/core/src/change_detection/constants.ts) ).

**Comment by alfaproject** [comment]
If you export as const then it won’t be in the bundle and can’t be used outside this repo anyway, right?

**Comment by QuentinFchx** [comment]
@alfaproject this it what I thought at first, but now I'm a little confused haha 😄.

**Comment by QuentinFchx** [comment]
Hello :)
Would it be possible to have a status on this?
If you think the PR is still relevant, I'll fix conflicts and hopefully have this merged 😉

**Comment by QuentinFchx** [comment]
I signed the CLA with my new email address!

**Comment by QuentinFchx** [comment]
Better late than never 😀

**Comment by QuentinFchx** [comment]
Hello @ahnpnl !
Would you have some time to review this by any chance? :bowtie:

**Comment by ahnpnl** [comment]
Hi all LGTM, just wonder if 103 is counted because I saw it on MDN.

**Comment by petebacondarwin** [comment]
@QuentinFchx - sorry that this has been around for so long. Could you rebase and then I will begin the review process.

**Comment by QuentinFchx** [comment]
No worries, I can do that.

Just to avoid some back and forth, in the current state of the PR I replaced all the "hardcoded" status codes in the tests by their equivalent in the enum.
Is it something we want to keep? It probably means we want some rules to avoid contributors from hardcoding them in the future (when writing new tests) and end up with 2 ways of having status code represented in the code.
However, I'm not sure I'll have the time to dig into that (how to implement such rules).

I'll do two separate commits:

- 1 with the enum + export + some documentation where relevant
- 1 with the status code replaced in the tests

and I'll let you decide if you want to keep both ;p

**Comment by petebacondarwin** [comment]
Two commits sounds good. Thanks

**Comment by QuentinFchx** [comment]
Here we go, I pushed the rebase.

I noticed there was already status codes within that repository [here](https://github.com/angular/angular/blob/master/packages/misc/angular-in-memory-web-api/src/http-status-codes.ts) (that were added since I opened the PR).

Also some hard-coded statuses can be found in `/aio` and `packages/service-worker`. I did not update these paths.

**Comment by petebacondarwin** [comment]
I'm afraid you will need to make friends with our lint check job and make your commit messages a bit longer :-)
Also I think the second commit is actually `test(http): ...`

**Comment by petebacondarwin** [comment]
For reference: https://pr23548-5e70f7f.ngbuilds.io/api/common/http/HttpStatusCode

**Review by petebacondarwin — APPROVED** [review]
LGTM - note that since these are defined by `const enum` they have zero runtime performance impact over just using constant literals.

Reviewed-for: public-api

**Comment by petebacondarwin** [comment]
Oh and you need to update the golden public API files by running:

```sh
yarn bazel run //packages/common:common_api.accept
```

File: packages/common/http/src/response.ts [hunk 2]
```diff
@@ -348,3 +348,78 @@ export class HttpErrorResponse extends HttpResponseBase implements Error {
...
+  Conflict = 409,
+  Gone = 410,
+  LengthRequired = 411,
+  PreconditionFailed = 412,
+  PayloadTooLarge = 413,
+  UriTooLong = 414,
+  UnsupportedMediaType = 415,
+  RangeNotSatisfiable = 416,
+  ExpectationFailed = 417,
+  ImATeapot = 418,
```
Comment (alxhub) [inline packages/common/http/src/response.ts:402]: This is the only one here that really matters.

**Review by mhevery — APPROVED** [review]

**Review by mhevery — APPROVED** [review]
reviewed-for: global-approvers

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:354201370:BASE:354201390:1611794616632:a77301b6)

**Pull request closed without merging**